### PR TITLE
[helm/elastic-agent] upgrade helm chart to collect rotated logs

### LINKED ISSUE #10889: Helm charts allow to ingest container rotated logs
**Describe the enhancement:**

Now that [filebeat supports ingesting GZIP files](https://www.elastic.co/docs/reference/beats/filebeat/filebeat-input-filestream#reading-gzip-files), update the Helm charts to allow to ingest container rotated logs, including GZIP compressed ones. 

**Describe a specific use case for the enhancement or feature:**

Ingesting rotated container logs

**What is the definition of done?**

 - using the helm chart the agent can be configured as opt-in to ingest rotated container logs, including gzip compressed logs
 - the relevant documentation is updated

#### Also, if possible, update the necessary agent docs to show how to ingest rotated container logs using 9.2.0. This means only docs changes, no changes on the helm chart or anywhere else that requires a release.

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

Upgates the elastic-agent helm chart to allow the collection of rotated logs, including the GZIP-compressed logs.

## Why is it important?

To allow the ingestion of rotated log files now filebeat can ingest GZIP-compressed files. 

## Checklist

- [x] I have read and understood the [pull request guidelines](https://github.com/elastic/elastic-agent/blob/main/CONTRIBUTING.md#pull-request-guidelines) of this project.
- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- ~~[ ] I have added an integration test or an E2E test~~

## Disruptive User Impact

When enabling the ingestion of rotated logs on an existing deployment, it cause a one-time re-ingestion of the logs.

The input ID when using autodiscover is now scoped by container name, instead of container id. It's done to avoid data duplication as explained below.

### Context: K8s logging architecture

Logs are written to:
```
/var/log/pods/<namespace>_<pod_name>_<pod_uid>/<container_name>/<restart_count>.log # Active log file
/var/log/pods/<namespace>_<pod_name>_<pod_uid>/<container_name>/<restart_count>.log.TIMRESTAMP` # 1st rotation, plain file
/var/log/pods/<namespace>_<pod_name>_<pod_uid>/<container_name>/<restart_count>.log.TIMRESTAMP.gz` # subsequenmtt rotations, gziped
```

Legacy symlinked path:
```
/var/log/containers/<pod_name>_<namespace>_<container_name>-<container_id>.log -> /var/log/pods/<namespace>_<pod_name>_<pod_uid>/<container_name>/<restart_count>.log
```

Thus during a container restart, a CrashLoopBackoff, it's possible to have 2
containers with the same name, but different IDs. The container which just crashed
and is being removed and the new container being created.

Just for the example, lets consider:

container name: foo
container id 1: id-1
container id 2: id-2

pod name: pod
namespace: ns
pod uid: uid

This leads to the following log files and symlinks during the transition:

```
/var/log/pods/ns_pod_uid/foo/0.log
/var/log/pods/ns_pod_uid/foo/1.log

/var/log/containers/pod_ns_foo-id-1.log -> /var/log/pods/ns_pod_uid/foo/0.log
/var/log/containers/pod_ns_foo-id-2.log -> /var/log/pods/ns_pod_uid/foo/1.log
```

### k8s log collection

1 filestream input per container. Input ID pattern: `kubernetes-container-logs-${data.kubernetes.pod.name}-${data.kubernetes.container.id}`

Thus, for each container id, one input is created.

### Rotated logs

Rotated logs are scoped by container name, and restart counter, not container ID. (_/var/log/pods/<namespace>_<pod_name>_<pod_uid>/**<container_name>/<restart_count>**.log_)
To collect rotated logs, the following wildcard would be used:
```
/var/log/pods/<namespace>_<pod_name>_<pod_uid>/<container_name>/*.log.*
```

```yaml
id: kubernetes-container-logs-${kubernetes.pod.name}-${kubernetes.container.id}
paths:
  - /var/log/containers/*<container_id>.log
  - /var/log/pods/<namespace>_<pod_name>_<pod_uid>/<container_name>/*.log.*
```

for the example above, during the container restart, there would be 2 inputs created:

```yaml
id: kubernetes-container-logs-pod-id-1
paths:
  - /var/log/containers/*id-1.log
  - /var/log/pods/ns_pod_uid/foo/*.log.*
```

```yaml
id: kubernetes-container-logs-pod-id-2
paths:
  - /var/log/containers/*id-2.log
  - /var/log/pods/ns_pod_uid/foo/*.log.*
```

Both inputs harvest `/var/log/pods/ns_pod_uid/foo/*.log.*`, duplicating the data.

Therefore, the ID pattern must be scoped by container name.

## How to test this PR locally
- start a elastic stack
- create a kind cluster
- deploy the agent using the current helm chart
    - on your deployment, go to `app/observabilityOnboarding/kubernetes/?category=kubernetes` and follow the steps. It'll be something like:
```shell
helm repo add elastic https://helm.elastic.co/ && \
helm install elastic-agent elastic/elastic-agent \
  --version 9.2.1 \
  -n kube-system \
  --set outputs.default.url=https:\/\/YOUR_DEPLOYMENT.elastic-cloud.com:443 \
  --set kubernetes.onboardingID=YOUR_ONBORDING_ID \
  --set kubernetes.enabled=true \
  --set outputs.default.type=ESPlainAuthAPI \
  --set outputs.default.api_key=$(echo "YOUR_API_KEY" | base64 -d)
```
- verify all is working, container logs are bein ingested
- deploy some flog containers
```shell
kubectl apply -f ./flog.yaml
```

<details>
<summary>
flog.yaml:
</summary>

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: flog-log-generator
spec:
  template:
    spec:
      containers:
        - name: flog-unstructured-cont-rot
          image: mingrammer/flog
          #          too small "-d" won't give kubelet time to rotate the files
          args: ["-t", "stdout", "-d", "1us", "-l"]
        - name: crashloop-1
          image: busybox
          imagePullPolicy: IfNotPresent
          command: ["sh", "-c", "echo 'Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Duis autem vel eum iriure dolor in hendrerit in vulputate velit esse molestie consequat, vel illum dolore eu feugiat nulla facilisis at vel illum dolore eu feugiat nulla facilisis at' >&2; sleep 1; exit 1"]
      restartPolicy: OnFailure
  backoffLimit: 10
```
</details>

- ensure the logs are being ingested. Use filter `log.file.path : *flog*`. Logs are ingested from `/var/log/containers/`
- package the helm chart
```shell
mage helm:package
```
- monitor the k8s pods
```shell
watch kubectl get pods --all-namespaces
```
- upgrade the helm chart
```shell
helm upgrade elastic-agent elastic-agent-9.3.0-beta.tgz \
  -n kube-system \
  --set outputs.default.url=https:\/\/YOUR_DEPLOYMENT.elastic-cloud.com:443 \
  --set kubernetes.onboardingID=YOUR_ONBORDING_ID \
  --set kubernetes.enabled=true \
  --set outputs.default.type=ESPlainAuthAPI \
  --set outputs.default.api_key=$(echo "YOUR_API_KEY" | base64 -d)
```
- check all the `agent-pernode-elastic-agent-*` pod is recreated, everything is
  normal, no configuration has changed, the logs are still coming from
  `/var/log/containers/`
  Screenshot from 2025-11-12 10-23-02
- check the agent config:
  - exec into the container
```shell
kubectl exec -it agent-pernode-elastic-agent-* -n kube-system -- /bin/bash
```
  - on the container, install some tools, grab a diagnostics and check it:
```shell
microdnf install -y less unzip
elastic-agent diagnostics
unzip -o -d /tmp/diag elastic-agent-diagnostics-*
less /tmp/diag/components/filestream-default/beat-rendered-config.yml
```
    - check the filestream input uses the old pattern:
```yaml
      id: kubernetes-container-logs-coredns-66bc5c9577-d7f2z-4e613cb4a69a7a3269711d92b0164fc37138f350a740148204eb91e7018dbe03
      index: logs-kubernetes.container_logs-default
      parsers:
        - container:
            format: auto
            stream: all
      paths:
        - /var/log/containers/*4e613cb4a69a7a3269711d92b0164fc37138f350a740148204eb91e7018dbe03.log
```
- update it to collect rotated logs:
```shell
helm upgrade elastic-agent elastic-agent-9.3.0-beta.tgz \
  -n kube-system \
  --set kubernetes.containers.logs.rotated_logs=true \
  --set outputs.default.url=https:\/\/YOUR_DEPLOYMENT.elastic-cloud.com:443 \
  --set kubernetes.onboardingID=YOUR_ONBORDING_ID \
  --set kubernetes.enabled=true \
  --set outputs.default.type=ESPlainAuthAPI \
  --set outputs.default.api_key=$(echo "YOUR_API_KEY" | base64 -d)
```
- check there is a spike in ingested logs and not the logs come from `/var/log/pods/`
  Screenshot from 2025-11-12 10-28-08

- check the rotated logs, plain and gzip, were ingested. Use the following filters `log.file.path : *flog*log.*` and `log.file.path : *flog*log.*.gz`:
  Screenshot from 2025-11-12 10-30-37
  Screenshot from 2025-11-12 10-28-24
 - check the agent config again:
- exec into the container
```shell
kubectl exec -it agent-pernode-elastic-agent-* -n kube-system -- /bin/bash
```
- on the container, install some tools, grab a diagnostics and check it:
```shell
microdnf install -y less unzip
elastic-agent diagnostics
unzip -o -d /tmp/diag elastic-agent-diagnostics-*
less /tmp/diag/components/filestream-default/beat-rendered-config.yml
```
    - check the filestream input uses the new pattern:
```yaml
      gzip_experimental: true
      id: kubernetes-container-logs-5906df5a-7185-4fc3-a2ec-9f088c729cf0-flog-unstructured-cont-rot
      index: logs-kubernetes.container_logs-default
      parsers:
        - container:
            format: auto
            stream: all
      paths:
        - /var/log/pods/default_flog-log-generator-vkz64_5906df5a-7185-4fc3-a2ec-9f088c729cf0/flog-unstructured-cont-rot/*.log*
```

## Related issues

- Closes #10889

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Comment by AndersonQ** [comment]
> The change looks correct to me, but you need to add the new option to `values.schema.json` for validation. It would be nice to include that in the contributing doc you've added too.

thanks for spotting that!
 
> You should also set the backport label. I take it that this will only go into 9.3?

oh, backport to 9.2 you mean? Yeah, it's an open question. I think strictly speaking it's a new feature in the chart, thus it needs to wait the next minor. Is it how you're dealing with the chart versions or not?
If it's enabling an existing feature in the agent, would it go on a path release?

@cmacknz what do you think? should it go on the next patch release on 9.2?

**Comment by swiatekm** [comment]
> oh, backport to 9.2 you mean? 

If you don't want to backport at all, you should set the `backport-skip` label. I'd default to shipping this in 9.3 unless you have a good reason to backport to 9.2.

File: deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl [hunk 1]
```diff
@@ -17,12 +17,19 @@ Config input for container logs
     namespace: {{ .Values.kubernetes.namespace }}
   use_output: {{ .Values.kubernetes.output }}
   streams:
+  {{- if $.Values.kubernetes.containers.logs.rotated_logs }}
+  - id: kubernetes-container-logs-${kubernetes.pod.uid}-${kubernetes.container.name}
+    gzip_experimental: true
```
Comment (cmacknz) [inline deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl:22]: If this only goes to 9.3, won't this be unnecessary because it will be on by default there, correct?

File: deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl [hunk 2]
```diff
@@ -17,12 +17,19 @@ Config input for container logs
     namespace: {{ .Values.kubernetes.namespace }}
   use_output: {{ .Values.kubernetes.output }}
   streams:
+  {{- if $.Values.kubernetes.containers.logs.rotated_logs }}
+  - id: kubernetes-container-logs-${kubernetes.pod.uid}-${kubernetes.container.name}
```
Comment (cmacknz) [inline deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl:21]: This creates an input per container, since we are going to get data duplication anyway, is there a more efficient way to do this with a single filestream input instead? Or fewer inputs?

File: deploy/helm/elastic-agent/values.schema.json [hunk 3]
```diff
@@ -229,6 +229,10 @@
                                     "type": "boolean",
                                     "description": "Enable containers logs stream."
                                 },
+                                "rotated_logs": {
+                                    "type": "boolean",
+                                    "description": "BETA: enable rotated logs ingestion. WARNING: enabling it might cause date re-ingestion on exising deployments. [ref](https://www.elastic.co/docs/reference/integrations/kubernetes/container-logs#ingesting-rotated-container-logs)."
```
Comment (cmacknz) [inline deploy/helm/elastic-agent/values.schema.json]:
```suggestion
                                    "description": "BETA: enable rotated logs ingestion. WARNING: enabling it might cause date re-ingestion on existing deployments. [ref](https://www.elastic.co/docs/reference/integrations/kubernetes/container-logs#ingesting-rotated-container-logs)."
```

The exising/existing typo exists in several places besides this one.

**Comment by cmacknz** [comment]
This is a new feature and not a bug fix so there is no reason to put this into 9.2.2 and later.

Also we should really have some kind of test for this, at minimum that we ingest any logs at all. There are existing integration tests of the Helm chart in https://github.com/elastic/elastic-agent/blob/0d505841660baf7a9353937c65debb09b3f06d29/testing/integration/k8s/otel_helm_test.go#L31

File: deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl (hunk 1 above)
Comment (AndersonQ) [reply deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl:22]: I don't think so. For an user upgrading, it'd cause data reingestion. You're thinking about enabling gzip, right? This I believe is safe, but I'm not sure about rotated logs as it requires changing the input ID.

File: deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl (hunk 2 above)
Comment (AndersonQ) [reply deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl:21]: > is there a more efficient way to do this with a single filestream input instead?

That would be a complete change on how filestream is used. Think about "fewer", I can see 3 possibilities and impacts:
 - 1 input per pod. I'm not sure how autodiscover would add the container metadata. It could mix up things. In a pod with 2 containers, A and B, the container metadata could be for A, but be added to the logs of A and B. The metadata is added as a processor in the input.
 - 1 single input. Collect everything under `/var/log/pods/`. It'd rely on the `add_kubernets_metadata`, no container metadata would be available, unless we modify the processor
 - 1 input for the pods that should have the logs collected. I'd need to understand better how autodiscover works or discuss with someone that know it better to see if it can be done. I'm not sure if it works.

So, I don't right now see a simple, no impact for the users, solution to change the number of inputs we create.

File: deploy/helm/elastic-agent/values.schema.json (hunk 3 above)
Comment (AndersonQ) [reply deploy/helm/elastic-agent/values.schema.json]: fixed on https://github.com/elastic/elastic-agent/pull/11129/commits/14a96cb7e0ef833bff1bc03ffe048148e47deb92

File: deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl (hunk 2 above)
Comment (cmacknz) [reply deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl:21]: Thanks, resolving as out of scope for this change

File: deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl (hunk 1 above)
Comment (AndersonQ) [reply deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl:22]: @cmacknz are you fine letting it there?
besides filebeat isn't updated yet to have it enabled by default. So either this PR needs it or it'll be blocked until filebeat is updated

Comment (cmacknz) [reply deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl:22]: >  For an user upgrading, it'd cause data reingestion.

OK yes so we need this to still be opt-in but we should stop calling it experimental.

The non-experimental way to opt-in/opt-out should be to use `exclude_files: [".gz$"]` like we mostly already do in existing configurations.

**Comment by AndersonQ** [comment]
@cmacknz, @swiatekm, could you re-review it?

File: testing/integration/k8s/agent_helm_test.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"helm.sh/helm/v3/pkg/cli/values"
+
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+)
+
+func TestAgentKubeStackHelm(t *testing.T) {
```
Comment (swiatekm) [inline testing/integration/k8s/agent_helm_test.go]:
```suggestion
func TestKubernetesAgentHelmRotatedLogs(t *testing.T) {
```

`KubeStack` implies opentelemetry-kube-stack, which is a completely different, otel-centric deployment method. The test name should also mention that it's concerned with rotated logs.

File: testing/integration/k8s/agent_helm_test.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+// k8sStepCheckLogFilesIngested creates a test step that verifies rotated logs are ingested
+// by querying Elasticsearch and checking for files matching the provided regex patterns.
+func k8sStepCheckLogFilesIngested(
+	info *define.Info,
+	dsType, dataset, datastreamNamespace, wildcardPath string,
+	expectedFiles ...expectedLogFile,
+) k8sTestStep {
+	return func(t *testing.T, ctx context.Context, kCtx k8sContext, namespace string) {
+		require.EventuallyWithT(t, func(t *assert.CollectT) {
```
Comment (swiatekm) [inline testing/integration/k8s/agent_helm_test.go]:
```suggestion
		require.EventuallyWithT(t, func(collectT *assert.CollectT) {
```

This is dangerous shadowing imo, even if it works. If you do rename it, be careful to rename all the uses inside the test function.

File: testing/integration/k8s/common.go [hunk 6]
```diff
@@ -563,3 +565,71 @@ func queryK8sNamespaceDataStream(dsType, dataset, datastreamNamespace, k8snamesp
...
+	Aggregations struct {
+		FilesCount struct {
+			DocCountErrorUpperBound int `json:"doc_count_error_upper_bound"`
+			SumOtherDocCount        int `json:"sum_other_doc_count"`
+			Buckets                 []struct {
+				Key      string `json:"key"`
+				DocCount int    `json:"doc_count"`
+			} `json:"buckets"`
+		} `json:"files_count"`
+	} `json:"aggregations"`
```
Comment (swiatekm) [inline testing/integration/k8s/common.go:634]: If you want to move this to estools later, then `Aggregations` should be a map, shouldn't it? If you want this struct to be generic, it should have other fields too, or perhaps even use our official [typed API](https://github.com/elastic/go-elasticsearch/blob/main/typedapi/core/search/response.go). And if not, then you can delete most of it, the tests use only a few fields.

File: testing/integration/k8s/otel_helm_test.go [hunk 7]
```diff
@@ -165,32 +166,40 @@ func TestOtelKubeStackHelm(t *testing.T) {
...
+		require.NoError(t, err, "failed to helm values")
+	}
+	return helmValues
+}
+
 // k8sStepCheckRunningPods checks the status of the agent inside the pods returned by the selector
 func k8sStepCheckRunningPods(podLabelSelector string, expectedPodNumber int, containerName string) k8sTestStep {
 	return func(t *testing.T, ctx context.Context, kCtx k8sContext, namespace string) {
-		require.Eventually(t, func() bool {
+		require.EventuallyWithTf(t, func(t *assert.CollectT) {
```
Comment (swiatekm) [inline testing/integration/k8s/otel_helm_test.go]: Same point as above about the test fixture shadowing.

File: testing/integration/k8s/agent_helm_test.go (hunk 5 above)
Comment (AndersonQ) [reply testing/integration/k8s/agent_helm_test.go]: It was intentional, I wanted to avoid using the wrong one inside the condition function. I don't mind changing.

File: testing/integration/k8s/otel_helm_test.go (hunk 7 above)
Comment (AndersonQ) [reply testing/integration/k8s/otel_helm_test.go]: done: https://github.com/elastic/elastic-agent/pull/11129/commits/38a5e6563b6c9c19fe4c2a52b7fd316505048ad2

File: testing/integration/k8s/agent_helm_test.go (hunk 5 above)
Comment (AndersonQ) [reply testing/integration/k8s/agent_helm_test.go]: done: https://github.com/elastic/elastic-agent/pull/11129/commits/38a5e6563b6c9c19fe4c2a52b7fd316505048ad2

File: testing/integration/k8s/agent_helm_test.go (hunk 4 above)
Comment (AndersonQ) [reply testing/integration/k8s/agent_helm_test.go]: done: https://github.com/elastic/elastic-agent/pull/11129/commits/38a5e6563b6c9c19fe4c2a52b7fd316505048ad2

File: deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl (hunk 1 above)
Comment (cmacknz) [reply deploy/helm/elastic-agent/templates/integrations/_kubernetes/_kubernetes_logs_containers.tpl:22]: If we are planning to leave this disabled by default this is fine, but we will need a plan for how to enable by default safely.

**Comment by AndersonQ** [comment]
@cmacknz is it good to go?

**Comment by cmacknz** [comment]
Yes Mikolaj's approval is enough here, I am confident he knows more about Helm than I do.

I do think we need a plan to get rid of the `gzip_experimental` flag as a follow up.

**Comment by AndersonQ** [comment]
> I do think we need a plan to get rid of the `gzip_experimental` flag as a follow up.

we have here: 
 - https://github.com/elastic/elastic-agent/issues/11559
